### ISSUE TITLE
Multiple ICMP requests get muddled when running concurrently

### ISSUE THREAD

**Issue by austin-barrington** [go-ping/ping#54]
**Problem:**
  Non pingable IP addresses are receiving ping results from other ping requests when run in parallel.

**Example Code:**
```
package main

import (
	"fmt"
	"sync"

	"github.com/sparrc/go-ping"
)

// PingCheck --
func PingCheck(HostToPing string) ping.Statistics {
	pinger, err := ping.NewPinger(HostToPing)
	if err != nil {
		panic(err)
	}

	pinger.Count = 1
	pinger.Timeout = 1000000000

	pinger.OnRecv = func(pkt *ping.Packet) {
		fmt.Printf("%d bytes from %s: icmp_seq=%d time=%v\n",
			pkt.Nbytes, pkt.IPAddr, pkt.Seq, pkt.Rtt)
	}

	pinger.OnFinish = func(stats ping.Statistics) {
		fmt.Printf("\n--- %s ping statistics ---\n", stats.Addr)
		fmt.Printf("%d packets transmitted, %d packets received, %v%% packet loss\n",
			stats.PacketsSent, stats.PacketsRecv, stats.PacketLoss)
		fmt.Printf("round-trip min/avg/max/stddev = %v/%v/%v/%v\n",
			stats.MinRtt, stats.AvgRtt, stats.MaxRtt, stats.StdDevRtt)
	}

	fmt.Printf("PING %s (%s):\n", pinger.Addr(), pinger.IPAddr())
	pinger.Run()                 // blocks until finished
	stats := pinger.Statistics() // get send/receive/rtt stats
	return stats
}

func main() {

	hosts := [...]string{"8.8.8.8", "1.2.3.4", "4.4.4.4"}

	sliceLength := len(hosts)
	var wg sync.WaitGroup
	wg.Add(sliceLength)
	for i := 0; i < sliceLength; i++ {
		go func(i int) {
			defer wg.Done()
			host := hosts[i]
			stats := PingCheck(host)
			fmt.Println(host, ": ", stats)
		}(i)
	}
	wg.Wait()
}

// 8.8.8.8 & 4.4.4.4 should work
// 1.2.3.4 should not but occasionally it returns with stats. This ip does not exist on the internet!

```

**Output:**
```
austinb-3GLM:agent austinb$ while true; do ./this && sleep 1; done | grep 1.2.3.4
PING 1.2.3.4 (1.2.3.4):
62 bytes from 1.2.3.4: icmp_seq=0 time=16.122ms
--- 1.2.3.4 ping statistics ---
1.2.3.4 :  {1 1 0 1.2.3.4 1.2.3.4 [16.122ms] 16.122ms 16.122ms 16.122ms 0s}
PING 1.2.3.4 (1.2.3.4):
--- 1.2.3.4 ping statistics ---
1.2.3.4 :  {0 1 100 1.2.3.4 1.2.3.4 [] 0s 0s 0s 0s}
PING 1.2.3.4 (1.2.3.4):
--- 1.2.3.4 ping statistics ---
1.2.3.4 :  {0 1 100 1.2.3.4 1.2.3.4 [] 0s 0s 0s 0s}
PING 1.2.3.4 (1.2.3.4):
--- 1.2.3.4 ping statistics ---
1.2.3.4 :  {0 1 100 1.2.3.4 1.2.3.4 [] 0s 0s 0s 0s}
PING 1.2.3.4 (1.2.3.4):
62 bytes from 1.2.3.4: icmp_seq=0 time=14.365ms
--- 1.2.3.4 ping statistics ---
1.2.3.4 :  {1 1 0 1.2.3.4 1.2.3.4 [14.365ms] 14.365ms 14.365ms 14.365ms 0s}
PING 1.2.3.4 (1.2.3.4):
62 bytes from 1.2.3.4: icmp_seq=0 time=14.198ms
--- 1.2.3.4 ping statistics ---
1.2.3.4 :  {1 1 0 1.2.3.4 1.2.3.4 [14.198ms] 14.198ms 14.198ms 14.198ms 0s}
PING 1.2.3.4 (1.2.3.4):
62 bytes from 1.2.3.4: icmp_seq=0 time=15.194ms
--- 1.2.3.4 ping statistics ---
1.2.3.4 :  {1 1 0 1.2.3.4 1.2.3.4 [15.194ms] 15.194ms 15.194ms 15.194ms 0s}
PING 1.2.3.4 (1.2.3.4):
--- 1.2.3.4 ping statistics ---
1.2.3.4 :  {0 1 100 1.2.3.4 1.2.3.4 [] 0s 0s 0s 0s}
```

**Comment by austin-barrington** [comment]
@sparrc Any ideas?

**Comment by sparrc** [comment]
Hi @austin-barrington, Might be a bug, but this part of your code looks a little suspect to me:

```go
	for i := 0; i < sliceLength; i++ {
		go func(i int) {
			defer wg.Done()
			host := hosts[i]
			stats := PingCheck(host)
			fmt.Println(host, ": ", stats)
		}(i)
	}
```

Because you are reusing the `i` variable within the gofunc. Could you change it to this and test again?:

```go
```go
	for i := 0; i < sliceLength; i++ {
		go func(j int) {
			defer wg.Done()
			host := hosts[j]
			stats := PingCheck(host)
			fmt.Println(host, ": ", stats)
		}(i)
	}
```
```

**Comment by wargebitebane-glich** [comment]
Hello @sparrc, your recommendation doesn't help.
**Output:**
PING 8.8.8.8 (8.8.8.8):
PING 4.4.4.4 (4.4.4.4):
PING 1.2.3.4 (1.2.3.4):
24 bytes from 4.4.4.4: icmp_seq=0 time=65.0315ms
24 bytes from 8.8.8.8: icmp_seq=0 time=65.0315ms
24 bytes from 1.2.3.4: icmp_seq=0 time=65.0315ms

--- 4.4.4.4 ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss       
round-trip min/avg/max/stddev = 65.0315ms/65.0315ms/65.0315ms/0s

--- 8.8.8.8 ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max/stddev = 65.0315ms/65.0315ms/65.0315ms/0s
8.8.8.8 :  &{1 1 0 8.8.8.8 8.8.8.8 [65.0315ms] 65.0315ms 65.0315ms 65.0315ms 0s}
4.4.4.4 :  &{1 1 0 4.4.4.4 4.4.4.4 [65.0315ms] 65.0315ms 65.0315ms 65.0315ms 0s}

--- 1.2.3.4 ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max/stddev = 65.0315ms/65.0315ms/65.0315ms/0s
1.2.3.4 :  &{1 1 0 1.2.3.4 1.2.3.4 [65.0315ms] 65.0315ms 65.0315ms 65.0315ms 0s}

**Comment by treydock** [comment]
Has there been any solution to this issue? I ran into this same problem.

**Comment by davidhoo** [comment]
> Has there been any solution to this issue? I ran into this same problem.

https://github.com/sparrc/go-ping/pull/84

**Comment by CHTJonas** [comment]
I'm failing to replicate this using the code in the most recent commit on `master`. Is anyone still experiencing this issue?

<details>
<summary>Code used to test</summary>

```go
package main

import (
        "fmt"
        "sync"

        "github.com/go-ping/ping"
)

func PingCheck(HostToPing string) *ping.Statistics {
        pinger, err := ping.NewPinger(HostToPing)
        if err != nil {
                panic(err)
        }

        pinger.Count = 10
        pinger.SetPrivileged(true)

        pinger.OnRecv = func(pkt *ping.Packet) {
                fmt.Printf("%d bytes from %s: icmp_seq=%d time=%v\n",
                        pkt.Nbytes, pkt.IPAddr, pkt.Seq, pkt.Rtt)
        }

        pinger.OnFinish = func(stats *ping.Statistics) {
                fmt.Printf("\n--- %s ping statistics ---\n", stats.Addr)
                fmt.Printf("%d packets transmitted, %d packets received, %v%% packet loss\n",
                        stats.PacketsSent, stats.PacketsRecv, stats.PacketLoss)
                fmt.Printf("round-trip min/avg/max/stddev = %v/%v/%v/%v\n",
                        stats.MinRtt, stats.AvgRtt, stats.MaxRtt, stats.StdDevRtt)
        }

        fmt.Printf("PING %s (%s):\n", pinger.Addr(), pinger.IPAddr())
        pinger.Run()
        return pinger.Statistics()
}

func main() {
        hosts := [...]string{"1.1.1.1", "8.8.8.8", "1.2.3.4", "9.9.9.9"}

        sliceLength := len(hosts)
        var wg sync.WaitGroup
        wg.Add(sliceLength)
        for i := 0; i < sliceLength; i++ {
                go func(j int) {
                        defer wg.Done()
                        host := hosts[j]
                        stats := PingCheck(host)
                        fmt.Println(host, ": ", stats)
                }(i)
        }
        wg.Wait()
}
```

</details>

**Comment by CHTJonas** [comment]
Should now be resolved by #130, #132 and #134.
